### PR TITLE
ARROW-8288: [Python] Expose with_ modifiers on DataType

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

File: python/pyarrow/types.pxi [hunk 1]
```diff
@@ -918,6 +918,69 @@ cdef class Field:
...
+        Returns
+        -------
+        field: pyarrow.Field
+        """
+        cdef:
+            shared_ptr[CField] field
+            c_bool c_nullable
+
+        c_nullable = bool(nullable)
+        with nogil:
```
Comment (jorisvandenbossche) [inline python/pyarrow/types.pxi]: Just wondering, is it worth releasing the GIL in this case? (since I suppose this is a very cheap operation)

**Review by jorisvandenbossche — APPROVED** [review]
Looks good to me

File: python/pyarrow/types.pxi (hunk 1 above)
Comment (xhochy) [reply python/pyarrow/types.pxi]: Not really sure if there is any benefit. As we are doing this also in other actions of this kind in the codebase, I also did it here for consistency.

Comment (jorisvandenbossche) [reply python/pyarrow/types.pxi]: (now, I also don't know to what extent it gives much overhead to release/re-acquire the gil for tiny operations)

**Pull request closed without merging**